src/tools/perf/lib/uct_tests.cc

```diff
@@ -169,8 +169,15 @@ class uct_perf_test_runner {
     static ucs_status_t am_hander(void *arg, void *data, size_t length,
                                   unsigned flags)
     {
-        ucs_assert(UCS_CIRCULAR_COMPARE8(*(psn_t*)arg, <=, *(psn_t*)data));
-        *(psn_t*)arg = *(psn_t*)data;
+        uct_perf_test_runner *self = (uct_perf_test_runner *)arg;
+
+        ucs_assert(UCS_CIRCULAR_COMPARE8(self->get_sn(self->m_perf.recv_buffer),
+                                         <=, self->get_sn(data)));
+
+        self->set_sn(self->m_perf.recv_buffer,
+                     self->m_perf.uct.recv_mem.mem_type,
+                     data, self->m_perf.uct.access_mem_type);
```

Comment:
if we copy it here to a recv buffer, which could be non host memory, it means need another copy later to check the sn?
maybe copy to some host buffer directly?
also 'data' is always host memory, it's not 'access_mem_type' - so it means set_sn/get_sn functions should not receive 2 mem types

Reply:
> it means need another copy later to check the sn?

yep

>maybe copy to some host buffer directly?

good catch - removed excessive copy

> also 'data' is always host memory, it's not 'access_mem_type' - so it means set_sn/get_sn functions should not receive 2 mem types

fixed